File: server/crypto.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
)

const (
	nonceSize = 12 // AES-GCM standard nonce
	keySize   = 32 // AES-256
	hkdfSalt  = "push-tunnel-v1"
)

// Crypto handles AES-256-GCM encryption with HKDF-derived keys.
type Crypto struct {
	aead cipher.AEAD
	key  []byte
}

// NewCrypto derives an AES-256 key from the PSK using HKDF-SHA256 and
// returns a ready-to-use Crypto instance.
func NewCrypto(psk string) (*Crypto, error) {
	hk := hkdf.New(sha256.New, []byte(psk), []byte(hkdfSalt), nil)
	key := make([]byte, keySize)
	if _, err := io.ReadFull(hk, key); err != nil {
		return nil, fmt.Errorf("hkdf: %w", err)
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return &Crypto{aead: aead, key: key}, nil
}

// Encrypt encrypts plaintext and returns base64(nonce || ciphertext || tag).
func (c *Crypto) Encrypt(plaintext []byte) (string, error) {
	nonce := make([]byte, nonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}
	ciphertext := c.aead.Seal(nonce, nonce, plaintext, nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt decodes base64 input and decrypts it.
func (c *Crypto) Decrypt(encoded string) ([]byte, error) {
	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fmt.Errorf("base64 decode: %w", err)
	}
	if len(data) < nonceSize+c.aead.Overhead() {
		return nil, errors.New("ciphertext too short")
	}
	nonce := data[:nonceSize]
	ciphertext := data[nonceSize:]
	plaintext, err := c.aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}
	return plaintext, nil
}

// ComputeAuthToken generates HMAC-SHA256(deviceID, timestamp) for request auth.
func ComputeAuthToken(deviceID string, timestamp string, key []byte) string {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(deviceID + ":" + timestamp))
	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
}

// VerifyAuthToken checks HMAC-SHA256 auth token.
func VerifyAuthToken(deviceID, timestamp, token string, key []byte) bool {
	expected := ComputeAuthToken(deviceID, timestamp, key)
	return hmac.Equal([]byte(expected), []byte(token))
}

File: client/lib/crypto.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

const String _hkdfSalt = 'push-tunnel-v1';
const int _keySize = 32;
const int _nonceSize = 12;

/// AES-256-GCM encryption matching the Go server implementation.
class TunnelCrypto {
  late final SecretKey _key;
  final AesGcm _algo = AesGcm.with256bits();

  TunnelCrypto._();

  /// Derive key from PSK using HKDF-SHA256 and initialise.
  static Future<TunnelCrypto> create(String psk) async {
    final c = TunnelCrypto._();
    final hkdf = Hkdf(hmac: Hmac(Sha256()), outputLength: _keySize);
    final ikm = SecretKey(utf8.encode(psk));
    final derived = await hkdf.deriveKey(
      secretKey: ikm,
      nonce: utf8.encode(_hkdfSalt),
      info: <int>[],
    );
    c._key = derived;
    return c;
  }

  /// Encrypt plaintext → base64(nonce[12] || ciphertext || tag[16]).
  Future<String> encrypt(Uint8List plaintext) async {
    final nonce = _randomBytes(_nonceSize);
    final secretBox = await _algo.encrypt(
      plaintext,
      secretKey: _key,
      nonce: nonce,
    );
    // secretBox.concatenation() returns nonce + ciphertext + mac
    final combined = secretBox.concatenation();
    return base64.encode(combined);
  }

  /// Decrypt base64-encoded ciphertext.
  Future<Uint8List> decrypt(String encoded) async {
    final data = base64.decode(encoded);
    if (data.length < _nonceSize + 16) {
      throw FormatException('Ciphertext too short');
    }
    final secretBox = SecretBox.fromConcatenation(
      data,
      nonceLength: _nonceSize,
      macLength: 16,
    );
    final plaintext = await _algo.decrypt(secretBox, secretKey: _key);
    return Uint8List.fromList(plaintext);
  }

  /// Compute HMAC-SHA256 auth token matching server's ComputeAuthToken.
  static Future<String> computeAuthToken(
      String deviceId, String timestamp, List<int> key) async {
    final hmac = Hmac(Sha256());
    final mac = await hmac.calculateMac(
      utf8.encode('$deviceId:$timestamp'),
      secretKey: SecretKey(key),
    );
    return base64.encode(mac.bytes);
  }

  /// Expose raw key bytes for auth token computation.
  Future<List<int>> get keyBytes async {
    return await _key.extractBytes();
  }

  static Uint8List _randomBytes(int n) {
    final rng = Random.secure();
    return Uint8List.fromList(List.generate(n, (_) => rng.nextInt(256)));
  }
}
